test/support/producer.ex

```diff
@@ -0,0 +1,62 @@
+defmodule Etl.Support.Producer do
+  defstruct pid: nil, partitions: [], hash: nil
+
+  defmodule Stage do
+    use GenStage
+
+    def start_link(opts) do
+      GenStage.start_link(__MODULE__, opts)
+    end
+
+    def init(t) do
+      case t.partitions do
+        [] ->
+          {:producer, t}
+
+        partitions when is_integer(partitions) ->
+          hash = t.hash || fn event -> {event, :erlang.phash2(event.data, partitions)} end
+          opts = [partitions: partitions, hash: hash]
+          {:producer, t, dispatcher: {GenStage.PartitionDispatcher, opts}}
+
+        partitions ->
+          hash = t.hash || fn event -> {event, :erlang.phash2(event, Enum.count(partitions))} end
```

Comment:
did you want this one to hash on the whole event or should it also hash on the event's `data` field?

Reply:
that is basically the default hashing algorithm gen_stage uses, but this is also in a test module.

Reply:
I just meant in the example above your hashing based on the `event.data` and here you’re using `event`. If that’s on purpose, cool but if not it could result in inconsistent hashing behavior that could be difficult to debug if the tests started failing later on.